src/api/icons.rs

```diff
@@ -650,7 +671,37 @@ fn save_icon(path: &str, icon: &[u8]) {
             create_dir_all(&CONFIG.icon_cache_folder()).expect("Error creating icon cache");
         }
         Err(e) => {
-            info!("Icon save error: {:?}", e);
+            warn!("Icon save error: {:?}", e);
         }
     }
 }
+
+fn get_icon_type(bytes: &[u8]) -> Option<&'static str> {
+    const PNG_HEADER: [u8; 4] = [137, 80, 78, 71];
+    const ICO_HEADER: [u8; 4] = [0, 0, 1, 0];
+    const WEBP_HEADER: [u8; 4] = [82, 73, 70, 70];
+    const JPG_HEADER: [u8; 3] = [255, 216, 255];
+    const BMP_HEADER: [u8; 2] = [66, 77];
+
+    if bytes.len() >= 4 {
+        let source_4bytes: Vec<u8> = bytes.iter().take(4).copied().collect();
+        let source_3bytes: Vec<u8> = source_4bytes.iter().take(3).copied().collect();
+        let source_2bytes: Vec<u8> = source_3bytes.iter().take(2).copied().collect();
+
+        return if source_4bytes == PNG_HEADER.to_vec() {
+            Some("png")
+        } else if source_4bytes == ICO_HEADER.to_vec() {
+            Some("x-icon")
+        } else if source_4bytes == WEBP_HEADER.to_vec() {
+            Some("webp")
+        } else if source_3bytes == JPG_HEADER.to_vec() {
+            Some("jpeg")
+        } else if source_2bytes == BMP_HEADER.to_vec() {
+            Some("bmp")
+        } else {
+            None
+        }
+    }
+
+    None
+}
```

Comment:
This could use [subslice patterns](https://blog.rust-lang.org/2020/03/12/Rust-1.42.html#subslice-patterns)

```suggestion
fn get_icon_type(bytes: &[u8]) -> Option<&'static str> {
    match bytes {
        [137, 80, 78, 71, ..] => Some("png"),
        [0, 0, 1, 0, ..] => Some("x-icon"),
        [82, 73, 70, 70, ..] => Some("webp"),
        [255, 216, 255, ..] => Some("jpeg"),
        [66, 77, ..] => Some("bmp"),
        _ => None
    }
}
```

Reply:
I tried that the hard way and it didn't worked, thx for the input. Will change it in a few!

Reply:
Updated code. Thanks for the help.